### PR TITLE
Feature/12 map photo popup

### PR DESCRIPTION
**Pull request by dmakarevich**

### CODE REVIEW

File: PhotoMapTests/Storyboards/StoryboardsTests.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//  StoryboardsTests.swift
+//  PhotoMapTests
+//
+//  Created by wsa-141-31b on 5/20/21.
+//
+
+import XCTest
+@testable import PhotoMap
+
+class StoryboardsTests: XCTestCase {
```
Comment (atereshkov) [inline PhotoMapTests/Storyboards/StoryboardsTests.swift:11]: Good catch

File: PhotoMapTests/ViewModels/MapPhotoViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        cancelBag = CancelBag()
+        diContainer = DIContainerMock()
+        coordinator = MapPhotoCoordinator(diContainer: diContainer)
+        viewModel = MapPhotoViewModel(coordinator: coordinator, diContainer: diContainer)
+    }
+
+    override func tearDownWithError() throws {
+        cancelBag = nil
+        diContainer = nil
+        viewModel = nil
```
Comment (atereshkov) [inline PhotoMapTests/ViewModels/MapPhotoViewModelTests.swift:28]: coordinator = nil

File: PhotoMapTests/ViewModels/MapViewModelTests.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        cancelBag = CancelBag()
+        diContainer = DIContainerMock()
+        coordinator = MapCoordinator(diContainer: diContainer)
+        viewModel = MapViewModel(coordinator: coordinator, diContainer: diContainer)
+    }
+
+    override func tearDownWithError() throws {
+        cancelBag = nil
+        diContainer = nil
+        viewModel = nil
```
Comment (atereshkov) [inline PhotoMapTests/ViewModels/MapViewModelTests.swift:28]: coordinator = nil

File: PhotoMap/Features/Map/ViewModels/MapViewModelType.swift [hunk 4]
```diff
@@ -9,17 +9,18 @@ import MapKit
...
+    var navigationButtonSubject: PassthroughSubject<UIControl, Never> { get }
+    var photoButtonSubject: PassthroughSubject<UIControl, Never> { get }
 }
 
 protocol MapViewModelOutput {
     var tabTitle: String { get }
     var isShowUserLocation: Bool { get }
     var region: MKCoordinateRegion? { get }
     var modeButtonCollor: UIColor { get }
+    var isFollowModeOn: Bool { get }
```
Comment (atereshkov) [inline PhotoMap/Features/Map/ViewModels/MapViewModelType.swift:23]: Add tests for other outputs (tabTitle, isShowUserLocation, region, modeButtonCollor)

File: PhotoMap/Features/MapPhoto/ViewModels/MapPhotoViewModelType.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    var cancelButtonSubject: PassthroughSubject<UIControl, Never> { get }
+    var categoryViewSubject: PassthroughSubject<GestureType, Never> { get }
+    var closeBarButtonSubject: PassthroughSubject<UIBarButtonItem, Never> { get }
+    var descriptionSubject: PassthroughSubject<String, Never> { get }
+    var doneButtonSubject: PassthroughSubject<UIControl, Never> { get }
+}
+
+protocol MapPhotoViewModelOutput {
+    var isHiddenCategoryPicker: Bool { get }
+    var categoryPublisher: Category? { get }
```
Comment (atereshkov) [inline PhotoMap/Features/MapPhoto/ViewModels/MapPhotoViewModelType.swift:21]: Add tests for categoryPublisher

File: PhotoMapTests/Mocks/Services/LocationServiceMock.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        super.init()
+
+        if #available(iOS 14.0, *) {
+            self.status.send(locationManager.authorizationStatus)
+        } else {
+            self.status.send(CLLocationManager.authorizationStatus())
+        }
+
+//        bind()
+        locationManager.startUpdatingLocation()
```
Comment (atereshkov) [inline PhotoMapTests/Mocks/Services/LocationServiceMock.swift:42]: No real logic should be placed in mocked/stubbed classes

**Review by atereshkov — CHANGES_REQUESTED** [review]
Improvements

File: PhotoMapTests/Mocks/Services/LocationServiceMock.swift (hunk 6 above)
Comment (dmakarevich) [reply PhotoMapTests/Mocks/Services/LocationServiceMock.swift]: fix 6f58eb2

File: PhotoMap/Features/MapPhoto/ViewModels/MapPhotoViewModelType.swift (hunk 5 above)
Comment (dmakarevich) [reply PhotoMap/Features/MapPhoto/ViewModels/MapPhotoViewModelType.swift:21]: fix 6f58eb2

File: PhotoMap/Features/Map/ViewModels/MapViewModelType.swift (hunk 4 above)
Comment (dmakarevich) [reply PhotoMap/Features/Map/ViewModels/MapViewModelType.swift:23]: fix 6f58eb2

File: PhotoMapTests/ViewModels/MapViewModelTests.swift (hunk 3 above)
Comment (dmakarevich) [reply PhotoMapTests/ViewModels/MapViewModelTests.swift:30]: fix 6f58eb2

File: PhotoMapTests/ViewModels/MapPhotoViewModelTests.swift (hunk 2 above)
Comment (dmakarevich) [reply PhotoMapTests/ViewModels/MapPhotoViewModelTests.swift:28]: fix 6f58eb2

**Review by atereshkov — APPROVED** [review]

**Pull request merged**